### PR TITLE
[QNN EP] Fix index-out-of-bounds bug in Slice builder when initializer is shared

### PR DESCRIPTION
**Pull request by adrianlizarraga**
### Description
There's an index-out-of-bounds bug that is triggered when a Slice operator shares an initializer with another operator that is processed first. In this case, QNN EP fails to properly initialize a `raw_starts` (or `raw_ends`) vector, which is later indexed by a call to `SliceOp::PrepareForComputeHelper()`.

### Motivation and Context
Fix bug that blocks https://github.com/microsoft/onnxruntime/pull/17764

### CODE REVIEW

File: onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc [hunk 1]
```diff
@@ -84,123 +123,68 @@ Status SliceOpBuilder::ProcessInputs(QnnModelWrapper& qnn_model_wrapper,
...
+    constexpr size_t axes_index = 3;
+    constexpr size_t steps_index = 4;
 
-  for (size_t input_i = 0; input_i < input_count; ++input_i) {
-    auto& input_name = inputs[input_i].node_arg.Name();
-    if (input_name.empty()) {
-      // Ignore unspecified/unused optional input
-      continue;
-    }
-    if (qnn_model_wrapper.IsQnnTensorWrapperExist(input_name)) {
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc:107]: Note: This early check caused the loop to continue and prevented the `raw_starts` or `raw_ends` data structures from being properly initialized. They should always be initialized even if another operator has already added them to the QNN model.

File: onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc [hunk 2]
```diff
@@ -84,123 +123,68 @@ Status SliceOpBuilder::ProcessInputs(QnnModelWrapper& qnn_model_wrapper,
...
     }
 
-    bool is_quantized_tensor = inputs[input_i].quant_param.has_value();
-    const auto* type_proto = inputs[input_i].node_arg.TypeAsProto();
-    ORT_RETURN_IF_ERROR(utils::GetQnnDataType(is_quantized_tensor, type_proto, qnn_data_type));
-
-    std::vector<uint32_t> input_shape;
-    ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[input_i].node_arg, input_shape), "Cannot get shape");
-
-    Qnn_QuantizeParams_t quantize_param = QNN_QUANTIZE_PARAMS_INIT;
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc:121]: Note: This `quantize_param` variable was not being used either.

File: onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc [hunk 3]
```diff
@@ -75,6 +72,45 @@ void SliceOpBuilder::GetDataFromAttribute(const NodeUnit& node_unit,
...
+
+  switch (data_type) {
+    case ONNX_NAMESPACE::TensorProto_DataType_INT64: {
+      const int64_t* tensor_data = reinterpret_cast<const int64_t*>(initializer_bytes.data());
+      size_t size = tensor_byte_size / sizeof(int64_t);
+      output.insert(output.end(), tensor_data, tensor_data + size);
+      break;
+    }
+    case ONNX_NAMESPACE::TensorProto_DataType_INT32: {
+      const int32_t* tensor_data = reinterpret_cast<const int32_t*>(initializer_bytes.data());
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc:100]: Is it gauranteed that data properly aligned and contigious?

File: onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc [hunk 4]
```diff
@@ -75,6 +72,45 @@ void SliceOpBuilder::GetDataFromAttribute(const NodeUnit& node_unit,
...
+
+  size_t tensor_byte_size = initializer_bytes.size();
+  const auto data_type = input_info.initializer_tensor->data_type();
+
+  Status status;
+
+  switch (data_type) {
+    case ONNX_NAMESPACE::TensorProto_DataType_INT64: {
+      const int64_t* tensor_data = reinterpret_cast<const int64_t*>(initializer_bytes.data());
+      size_t size = tensor_byte_size / sizeof(int64_t);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc:95]: Are we expecting to fetch a fixed amount of information? If so, perhaps we could do some checks?

**Review by yuslepukhin — DISMISSED** [review]

File: onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc [hunk 5]
```diff
@@ -84,123 +120,68 @@ Status SliceOpBuilder::ProcessInputs(QnnModelWrapper& qnn_model_wrapper,
...
+  ORT_RETURN_IF_ERROR(SliceOp::PrepareForComputeHelper(raw_starts, raw_ends, raw_axes, raw_steps, compute_metadata));
 
-Status SliceOpBuilder::ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_wrapper,
-                                                   const NodeUnit& node_unit,
-                                                   std::vector<std::string>&& input_names,
-                                                   const logging::Logger& logger,
-                                                   bool do_op_validation) const {
-  std::vector<uint32_t> ranges_dims{static_cast<uint32_t>(ranges_.size()), 3};
+  std::vector<uint32_t> ranges_dims{static_cast<uint32_t>(input_dimensions.size()), 3};
   std::vector<uint32_t> ranges_data;
```
[the remark this answers is not in the document]
Comment (adrianlizarraga) [reply onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc:180]: Added.

File: onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc (hunk 3 above)
Comment (adrianlizarraga) [reply onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc]: Yes, I added the following comment (above) to explain:

```C++
// UnpackInitializerData() uses ORT's protobuf utilities, which ensure that the initializer bytes are
// contiguous, aligned, and in the appropriate endianness. This is necessary to be able to reinterpret bytes
// as an array of larger elements.
```

File: onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc (hunk 4 above)
Comment (adrianlizarraga) [reply onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc]: Yes, we are expecting a fixed amount of information (based on number of bytes and size of each element). I created a helper function that reinterprets bytes to a span of larger elements in a way that is less error-prone.

File: onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc (hunk 3 above)
Comment (snnn) [reply onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc]: The `initializer_bytes` variable is just a normal vector, which uses the default malloc/free to manage memory.

Comment (adrianlizarraga) [reply onnxruntime/core/providers/qnn/builder/opbuilder/slice_op_builder.cc]: Yes, the contents of `initializer_bytes` are copied to a different output vector before it goes out of scope.

**Review by yuslepukhin — APPROVED** [review]

**Review by HectorSVC — APPROVED** [review]

**Pull request merged**